src/briefcase/integrations/docker.py

```diff
@@ -189,16 +200,131 @@ def _buildx_installed(cls, tools: ToolCache):
         except subprocess.CalledProcessError:
             raise BriefcaseCommandError(cls.BUILDX_PLUGIN_MISSING)
 
+    def _determine_docker_mode(self, image_tag: str | None = None):
+        """Determine Docker's operating mode from how it interacts with bind mounts.
+
+        Docker can be installed in different ways on Linux that significantly impact how
+        containers interact with the host system. Of particular note is ownership of
+        files and directories in bind mounts (i.e. mounts using --volume).
+
+        Traditionally, Docker would pass through the UID/GID of the user used inside a
+        container as the owner of files created within the bind mount. And since the
+        default user inside containers is root, the files would be owned by root on the
+        host file system; this prevents later interaction with those files by the host.
+        To work around this, the Dockerfile can use a step-down user with a UID and GID
+        that matches the host user running Docker.
+
+        Other installation methods of Docker, though, are not compatible with using such
+        a step-down user. This includes Docker Desktop and rootless Docker. In these
+        modes, Docker maps the host user to the root user inside the container; this
+        mapping is transparent and would require changes to the host environment to
+        disable, if it can be disabled at all. This allows files created in bind mounts
+        inside the container to be owned on the host file system by the user running
+        Docker. Additionally, though, because the host user is mapped to root inside the
+        container, any files that were created by the host user in the bind mount
+        outside the container are owned by root inside the container; therefore, a step-
+        down user could not interact with such bind mount files inside the container.
+
+        To accommodate these different modes, this checks which user owns a file that is
+        created inside a bind mount in the container. If the owning user of that file on
+        the host file system is root, then a step-down user is necessary inside
+        containers. If the owning user is the host user, root should be used.
+        """
+        write_test_filename = "container_write_test"
+        host_write_test_dir_path = Path.cwd()
```

Comment:
Using the `build` directory probably won't be possible here since this can run before it exists.

I considered using a temporary file....but it presents a few of its own questions I think.

For starters, we can't use Python's `tempfile` to create the file....because we need the file to be created inside the container.

However, we can use `tempfile.gettempdir()` as the host directory to mount in to the container perform the write test. First, this doesn't confirm a Docker container will have write permissions inside the project....but that'll be problematic no matter what. Second, this still leaves the possibility that we leave a `root`-owned file in the user's tmp directory....and to avoid issues with the file persisting from a previous Briefcase run, we'll need to effectively recreate a method to manufacture a random and unique filename.

These aren't insurmountable....but is why I went with this obvious approach. I can consider the temporary directory more.

Reply:
FWIW, I think creating the build directory if it doesn't exist would be reasonable. I can't think of any edge case where we'd end up with a build directory in a *weird* location - worst case is we get a build directory that is empty because it's only used to create a temporary permission check file. All the other locations that use the build directory are set up as "create if doesn't exist"; this would just be one more usage of that pattern.